### PR TITLE
LayerNorm using PyTorch

### PR DESCRIPTION
**Pull request by enpasos**
## Description ##

Similar to BatchNormalization there exist some other variants of normalizing data flowing through the network that have been implemented by the underlying ai frameworks. Here LayerNorm as one of them has been wired up to be used with PyTorch.

This PullRequest would close https://github.com/deepjavalibrary/djl/issues/1057.

### CODE REVIEW

File: api/src/main/java/ai/djl/nn/norm/LayerNorm.java [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+ * before the non-linearity. Unlike batch normalization, layer normalization performs exactly the
+ * same computation at training and test times. It is also straightforward to apply to recurrent
+ * neural networks by computing the normalization statistics separately at each time step. Layer
+ * normalization is very effective at stabilizing the hidden state dynamics in recurrent networks.
+ * Empirically, we show that layer normalization can substantially reduce the training time compared
+ * with previously published techniques."
+ */
+public class LayerNorm extends AbstractBlock {
+
+    private static final byte VERSION = 2;
```
Comment (frankfliu) [inline api/src/main/java/ai/djl/nn/norm/LayerNorm.java:59]: We can start from Version 1, this is used for serialize and deserialize the block

File: mxnet/mxnet-engine/src/main/java/ai/djl/mxnet/engine/MxNDArrayEx.java [hunk 2]
```diff
@@ -636,6 +636,12 @@ public NDList dropout(NDArray input, float rate, boolean training) {
         return getManager().invoke("_npx_dropout", new NDList(input), params);
     }
 
+    @Override
```
Comment (frankfliu) [inline mxnet/mxnet-engine/src/main/java/ai/djl/mxnet/engine/MxNDArrayEx.java:639]:
```suggestion
    /** {@inheritDoc} */
    @Override
```

File: pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java [hunk 3]
```diff
@@ -411,6 +411,18 @@ public NDList dropout(NDArray input, float rate, boolean training) {
         return new NDList(JniUtils.dropout((PtNDArray) input, rate, training));
     }
 
+    @Override
```
Comment (frankfliu) [inline pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java:414]:
```suggestion
    /** {@inheritDoc} */
    @Override
```

File: tensorflow/tensorflow-engine/src/main/java/ai/djl/tensorflow/engine/TfNDArrayEx.java [hunk 4]
```diff
@@ -378,6 +378,11 @@ public NDList dropout(NDArray input, float rate, boolean training) {
         throw new UnsupportedOperationException("Not implemented");
     }
 
+    @Override
```
Comment (frankfliu) [inline tensorflow/tensorflow-engine/src/main/java/ai/djl/tensorflow/engine/TfNDArrayEx.java:381]:
```suggestion
    /** {@inheritDoc} */
    @Override
```

File: pytorch/pytorch-native/src/main/native/ai_djl_pytorch_jni_PyTorchLibrary_nn_functional.cc [hunk 5]
```diff
@@ -144,6 +144,25 @@ JNIEXPORT jlong JNICALL Java_ai_djl_pytorch_jni_PyTorchLibrary_torchNNBatchNorm(
   API_END_RETURN()
 }
 
+JNIEXPORT jlong JNICALL Java_ai_djl_pytorch_jni_PyTorchLibrary_torchNNLayerNorm(JNIEnv* env, jobject jthis,
```
Comment (frankfliu) [inline pytorch/pytorch-native/src/main/native/ai_djl_pytorch_jni_PyTorchLibrary_nn_functional.cc:147]: Please run:
```
./gradlew formatCpp
```

File: api/src/main/java/ai/djl/nn/norm/LayerNorm.java [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+    public void prepare(Shape[] inputShapes) {
+        gamma.setShape(normalizedShape);
+        beta.setShape(normalizedShape);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    protected void saveMetadata(DataOutputStream os) throws IOException {
+        saveInputShapes(os);
+        os.writeInt(normalizedShape.getShape().length);
```
Comment (frankfliu) [inline api/src/main/java/ai/djl/nn/norm/LayerNorm.java:168]: This can be simplified as:
```
os.write(shape.getEncoded())
```

File: api/src/main/java/ai/djl/nn/norm/LayerNorm.java [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+        for (int i = 0; i < normalizedShape.getShape().length; i++) {
+            os.writeLong(normalizedShape.getShape()[i]);
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void loadMetadata(byte version, DataInputStream is)
+            throws IOException, MalformedModelException {
+        if (version == VERSION) {
```
Comment (frankfliu) [inline api/src/main/java/ai/djl/nn/norm/LayerNorm.java:178]: We don't have earlier version,
```
        if (version != VERSION) {
            throw new MalformedModelException("Unsupported encoding version: " + version);
        }
        readInputShapes(is);
```

File: api/src/main/java/ai/djl/nn/norm/LayerNorm.java [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+            readInputShapes(is);
+        } else if (version != 1) {
+            throw new MalformedModelException("Unsupported encoding version: " + version);
+        }
+        long[] shapeRaw = new long[is.readInt()];
```
Comment (frankfliu) [inline api/src/main/java/ai/djl/nn/norm/LayerNorm.java:183]:
```
normalizedShape = Shape.decode(is);
```

Comment (enpasos) [reply api/src/main/java/ai/djl/nn/norm/LayerNorm.java]: should be done now

File: api/src/main/java/ai/djl/nn/norm/LayerNorm.java (hunk 7 above)
Comment (enpasos) [reply api/src/main/java/ai/djl/nn/norm/LayerNorm.java]: fixed ... should be done now

File: api/src/main/java/ai/djl/nn/norm/LayerNorm.java (hunk 6 above)
Comment (enpasos) [reply api/src/main/java/ai/djl/nn/norm/LayerNorm.java]: fixed ... should be done now

File: api/src/main/java/ai/djl/nn/norm/LayerNorm.java (hunk 1 above)
Comment (enpasos) [reply api/src/main/java/ai/djl/nn/norm/LayerNorm.java]: fixed

File: pytorch/pytorch-native/src/main/native/ai_djl_pytorch_jni_PyTorchLibrary_nn_functional.cc (hunk 5 above)
Comment (enpasos) [reply pytorch/pytorch-native/src/main/native/ai_djl_pytorch_jni_PyTorchLibrary_nn_functional.cc:147]: I just did 

```
gradlew formatCpp
Found C:\djl\\gradle\wrapper\gradle-wrapper.jar

Deprecated Gradle features were used in this build, making it incompatible with Gradle 8.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/7.0.2/userguide/command_line_interface.html#sec:command_line_warnings

BUILD SUCCESSFUL in 1s
6 actionable tasks: 6 executed
```

but not seen an effect. Do I miss something?

Comment (frankfliu) [reply pytorch/pytorch-native/src/main/native/ai_djl_pytorch_jni_PyTorchLibrary_nn_functional.cc:147]: I see, formatCpp doesn't work on Windows
Here is what it should looks like:
```
JNIEXPORT jlong JNICALL Java_ai_djl_pytorch_jni_PyTorchLibrary_torchNNLayerNorm(
    JNIEnv* env, jobject jthis, jlong jinput, jlongArray jnormalizedshape, jlong jweight, jlong jbias, jdouble jeps) {
```

File: pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java [hunk 9]
```diff
@@ -411,6 +411,19 @@ public NDList dropout(NDArray input, float rate, boolean training) {
...
+            NDArray input, Shape normalizedShape, NDArray gamma, NDArray beta, float eps) {
+
+        return new NDList(
+                JniUtils.layerNorm(
+                        (PtNDArray) input,
+                        normalizedShape,
+                        (PtNDArray) gamma,
+                        (PtNDArray) beta,
+                        eps));
+    }
```
Comment (frankfliu) [inline pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java:426]: Add en empty line here

Comment (enpasos) [reply pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java:425]: I tryed to fix it ... but could be I am blind here ... I am using
`gradlew formatJava`
and
`gradlew build`

File: pytorch/pytorch-native/src/main/native/ai_djl_pytorch_jni_PyTorchLibrary_nn_functional.cc (hunk 5 above)
Comment (enpasos) [reply pytorch/pytorch-native/src/main/native/ai_djl_pytorch_jni_PyTorchLibrary_nn_functional.cc]: I copy/paste the two lines ... hope it fits ... looks like I should build djl on linux.

**Review by frankfliu — APPROVED** [review]

**Comment by frankfliu** [comment]
@lanking520 @stu1130 Please take a look.

**Review by stu1130 — APPROVED** [review]
